Add `Import#json`, `Export#json`, and `Export#email`

benjaminwil opened the pull request:
_This pull request closes #26._

This pull request allows the importing and exporting of JSON files. Previously only CSVs were supported. The interfaces for importing and exporting JSON have been designed to work the same way they already work for the CSV interfaces.

I've added documentation to the README for all of the JSON interfaces, so I won't go into more detail here. Refer to the changes to the README.

## Breaking changes

There is one breaking change here, which is that the `Export#email_csv` method has been renamed to `Export#email`. It takes a new `:format` argument which defaults to `:csv`. Users will need to update their application code accordingly:

```diff
-Export.email_csv(to: my_email)
+Export.email(to: my_email)
```

stevepolitodesign commented on .tool-versions:
We should probably bump the Ruby version, but I think we should do that as a separarte commit.

stevepolitodesign replied:
This is resolved in #28. You'll need to rebase off `main` to account for this, and fix CI.

stevepolitodesign commented on lib/art_vandelay.rb:
```diff
@@ -116,11 +146,11 @@ def model_name
+    def row(attributes, format: :array)
       if self.attributes.any?
-        filtered_values(attributes.slice(*standardized_attributes))
+        filtered_values(attributes.slice(*standardized_attributes), format: format)
```
[Minor]

We can actually omit the hash value thanks to [this feature](https://bugs.ruby-lang.org/issues/14579).

```suggestion
        filtered_values(attributes.slice(*standardized_attributes), format:)
```

benjaminwil replied:
Nice! I'm not sure I've ever seen this in the wild. 👀

stevepolitodesign commented on lib/art_vandelay.rb:
```diff
@@ -116,11 +146,11 @@ def model_name
+        filtered_values(attributes.slice(*standardized_attributes), format: format)
       else
-        filtered_values(attributes)
+        filtered_values(attributes, format: format)
```
[Minor]

Same here.

benjaminwil replied:
🙏

stevepolitodesign reviewed (commented):
Thank you for taking the time to address this! I left an initial review, but I think this is close.

Also noting, I will need to update #33 based on this change.

stevepolitodesign commented:
Also, don't forget to update [NEWS.md](https://github.com/thoughtbot/art_vandelay/blob/main/NEWS.md). Since this is the first unreleased feature, we can do something like this:


```md
Unreleased

Add relevant info here.

*Your Name*
```

stevepolitodesign commented:
Noting, I also need to fix CI.

benjaminwil commented:
I've rebased all of my commits to:

- resolve new linting errors reported by Standard on Ruby 3.3.4.
- resolve all the discussion comments on this PR.

And I've added a new commit with an entry in the NEWS file. I followed the format of the existing entries but added my name to the end like you suggested in your comment. Let me know if I need to revise.

benjaminwil commented:
CI is now passing almost everywhere, except on Ruby 3.0. It fails on the lint.

3.0 is now passed end-of-life, anyway, so maybe the solution is to remove it from the CI version matrix 🔪

benjaminwil commented:
Alternatively we could skip linting for 3.0!

stevepolitodesign commented:
> CI is now passing almost everywhere, except on Ruby 3.0. It fails on the lint.
> 
> 3.0 is now passed end-of-life, anyway, so maybe the solution is to remove it from the CI version matrix 🔪

@benjaminwil now that #36 is merged you are free to rebase 👍

benjaminwil commented:
Rebased. 👍

stevepolitodesign reviewed (approved):
This is great! I think there's an opportunity to further refactor this, but since this project is in its infancy, I'd rather get it out into the world.

Possible future improvements:

-  Creating `Art::Vandelay::Import.{from_csv, from_json}`
-  Creating `Art::Vandelay::Export.{from_csv, from_json}`
- Setting the `format` as an instance variable to limit the need to pass it through multiple methods
